Add Rinkeby testnet to docker development environment.

Ebonsignori opened the pull request:
Update README.md with instructions on how to setup environment and
add Agorex contract address to docker environment variable to be
read by python's config file.

Ebonsignori commented:
I can't figure out how to connect to the websocket from localhost (it's not an approved origin problem, I can't even get to the `@sio.on('connect')` lines in `websocket_server`.)

Everything else is working and the other services now get their information from a container running geth (rinkeby node) and providing a local rpc-http and rpc-ws provider. 

However, I don't see where in the code that the _websocket opens its ports for listening_. I've exposed port 8085 and set SERVERNAME to 0.0.0.0 to mimic what I saw in docker-compose.production, but to no avail. Any guidance on this subject, @JonathonDunford ?

JonathonDunford commented:
What IP is your websocket_server broadcasting on?

JonathonDunford commented:
@Ebonsignori 

It looks like this code may have been pulled from another PR as it has a bunch of code in it. We can either wait to pull it into master until the other PR goes through, or you can rebase it.

Ebonsignori commented:
Are you certain about that? I uncomment the ascii import and fixed the `if where not = ' '` syntax error because they were preventing me from testing containers. 

![image](https://user-images.githubusercontent.com/17055832/39615499-7bf2e9fc-4f44-11e8-99a9-3123756d792a.png)

JonathonDunford commented on README.md:
```diff
@@ -14,30 +14,53 @@ A good starting place to learn about the system is by taking a look at the sourc
+For Arch, the AUR contains geth as [`go-ethereum`](https://wiki.archlinux.org/index.php/Ethereum) and can easily be downloaded using yaourt.
+
+#### Step 2: Download Rinkeby node to computer
+Downloading the Rinkeby node will take about 10 minutes and a bit of hard drive space (approx. 5 gigs for me).
```
This entirely depends on network speed.

JonathonDunford commented on alembic/versions/47c6d9c6fd29_add_order_fill_column_to_orders_table.py:
```diff
@@ -0,0 +1,29 @@
+"""Add order fill column to Orders table
```
All of these alembic file changes are old.

You may need to rebase.

JonathonDunford commented on app/services/websocket_server.py:
```diff
@@ -162,31 +162,31 @@ def format_transfer(transfer):
         placeholder_args.append(Web3.toBytes(hexstr=token_give_hexstr))
         
     if token_get_hexstr:
-        if where not '':
```
This was all already changed in master. You definitely need a rebase.

JonathonDunford commented on docker-compose.development.yml:
```diff
@@ -4,15 +4,26 @@
+version: '2.3'
 services:
-  contract_observer:
+  geth:
```
Is there a way we could make running Geth through Docker optional?

We will probably want to consider running a separate parity node in a completely different server, so it would be nice to have an easy option in a config file or something to disable running of geth (and waiting for it)

Ebonsignori commented:
Added the option to use a remote node. 

Rebased and fixed what I could. 

Have no idea how alembic works, so I don't know what needs to be there.

JonathonDunford commented:
Did you git pull master before rebase? See the files that should be there:
https://github.com/Agorex-io/BackEnd/tree/master/alembic/versions

There are only 4 alembic files.

This PR is from Ebonsignori:add-testnet. Maybe your local master was out of date?
You'd need to pull Agorex/Master into your Ebonsignori/Master. Then, rebase add-testnet with the latest commit in that new master.

https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

Ebonsignori commented:
Yes, I fetched from upstream rather than origin where `git remote -v` returns this output:
```
endpoints	https://github.com/Agorex-io/BackEnd.git (fetch)
endpoints	https://github.com/Agorex-io/BackEnd.git (push)
origin	https://github.com/Ebonsignori/BackEnd.git (fetch)
origin	https://github.com/Ebonsignori/BackEnd.git (push)
upstream	https://github.com/Agorex-io/BackEnd.git (fetch)
upstream	https://github.com/Agorex-io/BackEnd.git (push)
```

My problem is that I **manually pulled in the old alembic files** because I couldn't get Docker to properly start with the changes you made to them. 

After re-adding the new alembic changes in master to this branch (simply replacing the entire directory), I get the same error as I did before when I run `docker-compose run contract_observer alembic upgrade head`:

```
[squirrel@taborlinthesquirrel BackEnd]$ docker-compose run contract_observer alembic upgrade head
Starting geth               ... done
Starting backend_postgres_1 ... done
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
/usr/local/lib/python3.6/site-packages/alembic/util/messaging.py:69: UserWarning: Revision 7719b4610d7e referenced from 7719b4610d7e -> e032ff832b82 (head), create ticker table is not present
  warnings.warn(msg)
Traceback (most recent call last):
  File "/usr/local/bin/alembic", line 11, in <module>
    load_entry_point('alembic==0.9.6', 'console_scripts', 'alembic')()
  File "/usr/local/lib/python3.6/site-packages/alembic/config.py", line 479, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/usr/local/lib/python3.6/site-packages/alembic/config.py", line 473, in main
    self.run_cmd(cfg, options)
  File "/usr/local/lib/python3.6/site-packages/alembic/config.py", line 456, in run_cmd
    **dict((k, getattr(options, k, None)) for k in kwarg)
  File "/usr/local/lib/python3.6/site-packages/alembic/command.py", line 254, in upgrade
    script.run_env()
  File "/usr/local/lib/python3.6/site-packages/alembic/script/base.py", line 425, in run_env
    util.load_python_file(self.dir, 'env.py')
  File "/usr/local/lib/python3.6/site-packages/alembic/util/pyfiles.py", line 81, in load_python_file
    module = load_module_py(module_id, path)
  File "/usr/local/lib/python3.6/site-packages/alembic/util/compat.py", line 83, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "alembic/env.py", line 70, in <module>
    run_migrations_online()
  File "alembic/env.py", line 65, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/usr/local/lib/python3.6/site-packages/alembic/runtime/environment.py", line 836, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/usr/local/lib/python3.6/site-packages/alembic/runtime/migration.py", line 321, in run_migrations
    for step in self._migrations_fn(heads, self):
  File "/usr/local/lib/python3.6/site-packages/alembic/command.py", line 243, in upgrade
    return script._upgrade_revs(revision, rev)
  File "/usr/local/lib/python3.6/site-packages/alembic/script/base.py", line 334, in _upgrade_revs
    revs = list(revs)
  File "/usr/local/lib/python3.6/site-packages/alembic/script/revision.py", line 656, in _iterate_revisions
    uppers = util.dedupe_tuple(self.get_revisions(upper))
  File "/usr/local/lib/python3.6/site-packages/alembic/script/revision.py", line 301, in get_revisions
    resolved_id, branch_label = self._resolve_revision_number(id_)
  File "/usr/local/lib/python3.6/site-packages/alembic/script/revision.py", line 437, in _resolve_revision_number
    self._revision_map
  File "/usr/local/lib/python3.6/site-packages/alembic/util/langhelpers.py", line 239, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/usr/local/lib/python3.6/site-packages/alembic/script/revision.py", line 152, in _revision_map
    down_revision = map_[downrev]
KeyError: '7719b4610d7e'
```
If I try to just run `docker-compose up` after this error, the database didn't properly initialize, so I get multiple null table call errors, like `asyncpg.exceptions.UndefinedTableError: relation "trades" does not exist`

**Have you successfully run the master branch in docker yet?** Because I can't on either windows or linux, however I was able to run my branch with the old alembic database config files.

Ebonsignori commented:
@JonathonDunford Yes, I fetched from upstream rather than origin where `git remote -v` returns this output:
```
endpoints	https://github.com/Agorex-io/BackEnd.git (fetch)
endpoints	https://github.com/Agorex-io/BackEnd.git (push)
origin	https://github.com/Ebonsignori/BackEnd.git (fetch)
origin	https://github.com/Ebonsignori/BackEnd.git (push)
upstream	https://github.com/Agorex-io/BackEnd.git (fetch)
upstream	https://github.com/Agorex-io/BackEnd.git (push)
```

What you're interpretting as a failed rebase is that I **manually pulled in the old alembic files** because I couldn't get Docker to properly start with the changes you made to them. 

After re-adding the new alembic changes in master to this branch (simply replacing the entire directory), I get the same error as I did before when I run `docker-compose run contract_observer alembic upgrade head`:

```
[squirrel@taborlinthesquirrel BackEnd]$ docker-compose run contract_observer alembic upgrade head
Starting geth               ... done
Starting backend_postgres_1 ... done
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
/usr/local/lib/python3.6/site-packages/alembic/util/messaging.py:69: UserWarning: Revision 7719b4610d7e referenced from 7719b4610d7e -> e032ff832b82 (head), create ticker table is not present
  warnings.warn(msg)
Traceback (most recent call last):
  File "/usr/local/bin/alembic", line 11, in <module>
    load_entry_point('alembic==0.9.6', 'console_scripts', 'alembic')()
  File "/usr/local/lib/python3.6/site-packages/alembic/config.py", line 479, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/usr/local/lib/python3.6/site-packages/alembic/config.py", line 473, in main
    self.run_cmd(cfg, options)
  File "/usr/local/lib/python3.6/site-packages/alembic/config.py", line 456, in run_cmd
    **dict((k, getattr(options, k, None)) for k in kwarg)
  File "/usr/local/lib/python3.6/site-packages/alembic/command.py", line 254, in upgrade
    script.run_env()
  File "/usr/local/lib/python3.6/site-packages/alembic/script/base.py", line 425, in run_env
    util.load_python_file(self.dir, 'env.py')
  File "/usr/local/lib/python3.6/site-packages/alembic/util/pyfiles.py", line 81, in load_python_file
    module = load_module_py(module_id, path)
  File "/usr/local/lib/python3.6/site-packages/alembic/util/compat.py", line 83, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "alembic/env.py", line 70, in <module>
    run_migrations_online()
  File "alembic/env.py", line 65, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/usr/local/lib/python3.6/site-packages/alembic/runtime/environment.py", line 836, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/usr/local/lib/python3.6/site-packages/alembic/runtime/migration.py", line 321, in run_migrations
    for step in self._migrations_fn(heads, self):
  File "/usr/local/lib/python3.6/site-packages/alembic/command.py", line 243, in upgrade
    return script._upgrade_revs(revision, rev)
  File "/usr/local/lib/python3.6/site-packages/alembic/script/base.py", line 334, in _upgrade_revs
    revs = list(revs)
  File "/usr/local/lib/python3.6/site-packages/alembic/script/revision.py", line 656, in _iterate_revisions
    uppers = util.dedupe_tuple(self.get_revisions(upper))
  File "/usr/local/lib/python3.6/site-packages/alembic/script/revision.py", line 301, in get_revisions
    resolved_id, branch_label = self._resolve_revision_number(id_)
  File "/usr/local/lib/python3.6/site-packages/alembic/script/revision.py", line 437, in _resolve_revision_number
    self._revision_map
  File "/usr/local/lib/python3.6/site-packages/alembic/util/langhelpers.py", line 239, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/usr/local/lib/python3.6/site-packages/alembic/script/revision.py", line 152, in _revision_map
    down_revision = map_[downrev]
KeyError: '7719b4610d7e'
```
If I try to just run `docker-compose up` after this error, the database didn't properly initialize, so I get multiple null table call errors, like `asyncpg.exceptions.UndefinedTableError: relation "trades" does not exist`

**Have you successfully run the master branch in docker yet?** Because I can't on either windows or linux, however I am able to run my branch with the old alembic database config files.

JonathonDunford commented:
@Ebonsignori 

https://github.com/Agorex-io/BackEnd/pull/56

Should fix that.

JonathonDunford commented:
@Ebonsignori 

Is this ready now?

Ebonsignori commented:
Yes, :shitpit:! Checkout the README for a short tutorial of how I provided dev environment options

JonathonDunford commented:
@Ebonsignori 

Where did you make the docker ran Geth optional?

I still see ports exposed and "depends on geth"

It'd be nice to have a simple config variable (or environment) that makes the backedn run geth through docker, but still allow the backend to point toa  remote parity/geth node instead.

Ebonsignori commented:
@JonathonDunford it's in the README.md

What you choose to uncomment in the `default.env` file that turns into / is copied to the `.env` file decides what node the backend points to. Uncommenting the first block allows docker-compose.development.yml to extend (overwrite part of) docker-compose.yml, allowing you to connect to a remote JSON-RPC provider (parity, geth) hosting a Rinkeby, Ethereum, or whatever node. Uncommenting the second block in `default.env` allows you to connect to a local geth providing JSON-RPC HTTP and WS for a Rinkeby node (this could be tweaked to be an Ethereum node if we want). 

The depends on and links shouldn't affect the other config overwrites, but it was sloppy to keep them in `docker-compose.yml`, so I fixed that with another commit.

JonathonDunford commented on app/config.py:
```diff
@@ -1,18 +1,24 @@
 POSTGRES_PASSWORD = environ.get("POSTGRES_PASSWORD")
 
-FRONTEND_CONFIG_FILE="https://raw.githubusercontent.com/agorex-io/FrontEnd/master/config/main.json"
+FRONTEND_CONFIG_FILE="https://raw.githubusercontent.com/Agorex-io/New-Frontend/master/src/js/constants/config.json"
```
Make this an environment variable.
